File: patcher/core/src/main/java/me/mamiiblt/instafel/patcher/core/patches/UnlockDeveloperOptions.java
package me.mamiiblt.instafel.patcher.core.patches;

import java.io.File;
import java.io.IOException;
import java.util.Iterator;
import java.util.List;

import me.mamiiblt.instafel.patcher.core.source.SmaliParser;
import me.mamiiblt.instafel.patcher.core.source.SmaliParser.SmaliInstruction;
import me.mamiiblt.instafel.patcher.core.utils.Log;
import me.mamiiblt.instafel.patcher.core.utils.SmaliUtils;
import me.mamiiblt.instafel.patcher.core.utils.Utils;
import me.mamiiblt.instafel.patcher.core.utils.models.LineData;
import me.mamiiblt.instafel.patcher.core.utils.patch.InstafelPatch;
import me.mamiiblt.instafel.patcher.core.utils.patch.InstafelTask;
import me.mamiiblt.instafel.patcher.core.utils.patch.PInfos;
import org.apache.commons.io.FileUtils;

@PInfos.PatchInfo(
    name = "Unlock Developer Options",
    shortname = "unlock_developer_options",
    desc = "You can unlock developer options with applying this patch!",
    author = "mamiiblt",
    isSingle = true
)
public class UnlockDeveloperOptions extends InstafelPatch {

    private SmaliUtils smaliUtils = getSmaliUtils();
    private String className = null;
    private File unlockRefSmali = null;

    @Override
    public List<InstafelTask> initializeTasks() {
        return List.of(
                findReferanceSmaliFile,
                getDevOptionsClass,
                addConstraintLineTask
        );
    }

    InstafelTask getDevOptionsClass = new InstafelTask("Get contraint defination class") {
        @Override
        public void execute() throws Exception {
            List<String> referanceFileContent = smaliUtils.getSmaliFileContent(unlockRefSmali.getAbsolutePath());
            List<LineData> linesWithInvokeAndUserSession = smaliUtils.getContainLines(
                    referanceFileContent, "(Lcom/instagram/common/session/UserSession;)Z", "invoke-static"
            );

            if (linesWithInvokeAndUserSession.size() != 1) {
                for (String line : referanceFileContent) {
                    Log.info(line );
                }
                failure("Static caller opcode can't found or more than 1!");
            }

            LineData callLine = linesWithInvokeAndUserSession.get(0);
            SmaliInstruction callLineInstruction = SmaliParser.parseInstruction(callLine.getContent(), callLine.getNum());
            className = callLineInstruction.getClassName().replace("LX/", "").replace(";", "");
            success("DevOptions class is " + className);
        }
    };

    InstafelTask findReferanceSmaliFile = new InstafelTask("Find referance smali file in X folders") {
        @Override
        public void execute() throws Exception {
            boolean fileFoundLock = false;

            for (File folder : smaliUtils.getSmaliFolders()) {
                if (fileFoundLock) {
                    break;
                } else {
                    File xFolder = new File(Utils.mergePaths(folder.getAbsolutePath(), "X"));
                    Log.info("Searching in X folder of " + folder.getName());

                    Iterator<File> fileIterator = FileUtils.iterateFiles(xFolder, null, true);
                    while (fileIterator.hasNext()) {
                        File file = fileIterator.next();
                        List<String> fContent = smaliUtils.getSmaliFileContent(file.getAbsolutePath());

                        boolean[] conditions = {false, false, false, false};
                        for (String line : fContent) {
                            if (line.contains(".field public final") && line.contains(":Lcom/google/common/collect/EvictingQueue;")) {
                                conditions[0] = true;
                            }

                            if (line.contains(".field public final") && line.contains(":Lcom/instagram/common/session/UserSession;")) {
                                conditions[1] = true;
                            }

                            if (line.contains(".field public") && line.contains(":Ljava/lang/String;")) {
                                conditions[2] = true;
                            }

                            if (line.contains(".super LX/")) {
                                conditions[3] = true;
                            }
                        }

                        boolean passStatus = true;
                        for (boolean cond : conditions) {
                            if (!cond) {
                                passStatus = false;
                            }
                        }

                        if (passStatus) {
                            unlockRefSmali = file;
                            Log.info("File found in " + unlockRefSmali.getName() + " at " + folder.getName());
                            fileFoundLock = true;
                            break;
                        }
                    }
                }
            }
        }
    };

    InstafelTask addConstraintLineTask = new InstafelTask("Add constraint line to DevOptions class") {
        @Override
        public void execute() throws IOException {
            File devOptionsFile = smaliUtils.getSmaliFilesByName(
                "X/" + className + ".smali"
            ).get(0);
            List<String> devOptionsContent = smaliUtils.getSmaliFileContent(devOptionsFile.getAbsolutePath());
            List<LineData> moveResultLines = smaliUtils.getContainLines(
                devOptionsContent, "move-result", "v0");
            if (moveResultLines.size() != 1) {
                failure("Move result line size is 0 or bigger than 1");
            } 
            LineData moveResultLine = moveResultLines.get(0);
            if (devOptionsContent.get(moveResultLine.getNum() + 2).contains("const v0, 0x1")) {
                failure("Developer options already unlocked.");
            }

            devOptionsContent.add(moveResultLine.getNum() + 1, "    ");
            devOptionsContent.add(moveResultLine.getNum() + 2, "    const v0, 0x1");
            smaliUtils.writeContentIntoFile(devOptionsFile.getAbsolutePath(), devOptionsContent);
            Log.info("Contraint added succesfully.");
            success("Developer options unlocked succesfully.");
        }
    };
}

File: patcher-core/src/main/kotlin/instafel/patcher/core/patches/UnlockDeveloperOptions.kt
/*
 * (c) 2026 Muhammed Ali Bulut, All rights reserved.
 *
 * See LICENSE file in repository root for copy file of license. For copyright
 * notices, technical issues, feedback, or any other related to this code file or
 * project, please contact me via mamii@mamii.dev or other ways.
 */

package instafel.patcher.core.patches

import instafel.patcher.core.source.SmaliParser
import instafel.patcher.core.utils.Log
import instafel.patcher.core.utils.modals.LineData
import instafel.patcher.core.utils.patch.InstafelPatch
import instafel.patcher.core.utils.patch.InstafelTask
import instafel.patcher.core.utils.patch.PInfos

@PInfos.PatchInfo(
    name = "Unlock Developer Options",
    shortname = "unlock_developer_options",
    desc = "You can unlock developer options with applying this patch!",
    isSingle = true
)
class UnlockDeveloperOptions: InstafelPatch() {

    lateinit var className: String

    // Fallback chain: try each reference class in order until one yields a
    // validated A00(UserSession)Z target. A reference is skipped (not fatal)
    // on ambiguous/missing file, no candidates, or zero valid candidates.
    private val referenceClasses = listOf(
        "/com/instagram/notifications/push/ClearNotificationReceiver.smali",
        "/com/instagram/base/activity/BaseFragmentActivity.smali",
        "/com/instagram/business/promote/activity/PromoteActivity.smali"
    )

    override fun initializeTasks() = mutableListOf(

        @PInfos.TaskInfo("Get constraint definition class")
        object: InstafelTask() {
            override fun execute() {

                var resolvedClassName: String? = null

                for (refPath in referenceClasses) {

                    val refResults = smaliUtils.getSmaliFilesByName(refPath)

                    if (refResults.isEmpty() || refResults.size > 1) {
                        Log.info("Reference $refPath ambiguous or missing (${refResults.size} matches), trying next.")
                        continue
                    }

                    val referenceFileContent =
                        smaliUtils.getSmaliFileContent(refResults.first().absolutePath)

                    val candidates = referenceFileContent.mapIndexedNotNull { index, line ->
                        val t = line.trim()
                        if (t.contains("invoke-static") &&
                            t.contains("->A00(") &&
                            t.contains("Lcom/instagram/common/session/UserSession;") &&
                            t.contains(")Z")
                        ) LineData(index, line) else null
                    }

                    if (candidates.isEmpty()) {
                        Log.info("No invoke-static A00(UserSession)Z found in $refPath, trying next reference.")
                        continue
                    }

                    var validLine: LineData? = null

                    for (lineData in candidates) {
                        val instruction = SmaliParser.parseInstruction(lineData.content, lineData.num)
                        val extractedClass = instruction.className.replace("LX/", "").replace(";", "")

                        val targetFile = smaliUtils.getSmaliFilesByName("X/$extractedClass.smali")
                            .firstOrNull() ?: continue

                        val targetContent = smaliUtils.getSmaliFileContent(targetFile.absolutePath)

                        val isValid = targetContent.any {
                            val t = it.trim()
                            t.contains(".method") &&
                            t.contains("A00") &&
                            t.contains("Lcom/instagram/common/session/UserSession;") &&
                            t.contains(")Z")
                        }

                        if (isValid) {
                            validLine = lineData
                            break
                        }
                    }

                    if (validLine == null) {
                        Log.info("No candidate validated in $refPath, trying next reference.")
                        continue
                    }

                    val callLineInstruction = SmaliParser.parseInstruction(validLine.content, validLine.num)
                    resolvedClassName = callLineInstruction.className.replace("LX/", "").replace(";", "")
                    break
                }

                if (resolvedClassName == null) {
                    failure("No valid DevOptions class found across any reference class.")
                    return
                }

                className = resolvedClassName
                success("DevOptions class is $className")
            }
        },

        @PInfos.TaskInfo("Add constraint line to DevOptions class")
        object: InstafelTask() {
            override fun execute() {

                val devOptionsFile = smaliUtils.getSmaliFilesByName("X/$className.smali")
                    .firstOrNull() ?: run {
                    failure("Developer options file not found")
                    return
                }

                val devOptionsContent =
                    smaliUtils.getSmaliFileContent(devOptionsFile.absolutePath).toMutableList()

                // Bound the search to the A00 method body — sibling methods (A01, etc.)
                // in the same file can share the exact same move-result/return shape.
                val methodStart = devOptionsContent.indexOfFirst { line ->
                    val t = line.trim()
                    t.contains(".method") &&
                    t.contains("A00") &&
                    t.contains("Lcom/instagram/common/session/UserSession;") &&
                    t.contains(")Z")
                }

                if (methodStart == -1) {
                    failure("A00(UserSession)Z method not found in $className")
                    return
                }

                val methodEnd = (methodStart until devOptionsContent.size)
                    .firstOrNull { devOptionsContent[it].trim() == ".end method" }

                if (methodEnd == null) {
                    failure("Could not find end of A00 method body")
                    return
                }

                val moveResultIndices = (methodStart..methodEnd)
                    .filter { devOptionsContent[it].trim() == "move-result v0" }

                if (moveResultIndices.size != 1) {
                    failure("Move result line count within A00 is ${moveResultIndices.size}, expected exactly 1.")
                    return
                }

                val moveResultIndex = moveResultIndices.first()

                if (devOptionsContent[moveResultIndex + 2].contains("const v0, 0x1")) {
                    Log.info("Developer options already unlocked.")
                    success("Developer options already unlocked, skipping.")
                    return
                }

                devOptionsContent.add(moveResultIndex + 1, "    ")
                devOptionsContent.add(moveResultIndex + 2, "    const v0, 0x1")

                smaliUtils.writeContentIntoFile(devOptionsFile.absolutePath, devOptionsContent)

                Log.info("Constraint added successfully.")
                success("Developer options unlocked successfully.")
            }
        }
    )
}